go/pools/fast_pool.go

```diff
@@ -0,0 +1,604 @@
+// It waits for all resources to be returned (Put).
+// After a Close, Get is not allowed.
+func (p *FastPool) Close() {
+	_ = p.SetCapacity(0, true)
+}
+
+// IsClosed returns true if the resource pool is closed.
+func (p *FastPool) IsClosed() bool {
+	return p.State().Closed
+}
+
+// Get will return the next available resource. If capacity
+// has not been reached, it will create a new one using the factory.
+// Otherwise, it will wait till the next resource becomes available or a timeout.
+func (p *FastPool) Get(ctx context.Context) (resource Resource, err error) {
+	if p.State().Closed {
+		return nil, ErrClosed
+	}
+
+	select {
+	case wrapper, ok := <-p.pool:
+		if !ok {
+			return nil, ErrClosed
+		}
+
+		p.Lock()
+		p.state.InPool--
+		p.state.InUse++
+		p.Unlock()
+
+		return wrapper.resource, nil
+
+	case <-ctx.Done():
+		return nil, ErrTimeout
+
+	default:
+		wrapper, err := p.safeCreate()
+		if err == errNeedToQueue {
+			return p.getQueued(ctx)
+		} else if err != nil {
+			return nil, err
+		}
+
+		return wrapper.resource, nil
+	}
+}
+
+// getQueued will wait for a resource to become available in the pool.
+// This is called when there is no capacity to create a resource and
+// there is nothing in the pool.
+func (p *FastPool) getQueued(ctx context.Context) (Resource, error) {
+	startTime := time.Now()
+
+	p.Lock()
+	p.state.Waiters++
+	p.Unlock()
+
+	for {
+		select {
+		case wrapper, ok := <-p.pool:
+			if !ok {
+				p.Lock()
+				p.state.Waiters--
+				p.Unlock()
+
+				return nil, ErrClosed
+			}
+
+			p.Lock()
+			p.state.InPool--
+			p.state.InUse++
+			p.state.Waiters--
+			p.state.WaitCount++
+			p.state.WaitTime += time.Now().Sub(startTime)
+			p.Unlock()
+
+			return wrapper.resource, nil
+
+		case <-ctx.Done():
+			p.Lock()
+			p.state.Waiters--
+			p.Unlock()
+
+			return nil, ErrTimeout
+
+		case <-time.After(100 * time.Millisecond):
+			// There could be a condition where this caller has been
+			// put into a queue, but another caller has failed in creating
+			// a resource, causing a deadlock. We'll check occasionally to see
+			// if there is now capacity to create.
+			if p.State().Closed {
+				return nil, ErrClosed
+			}
+			wrapper, err := p.safeCreate()
+			if err == errNeedToQueue {
+				continue
+			} else if err != nil {
+				return nil, err
+			} else {
+				return wrapper.resource, nil
+			}
+		}
+	}
+}
+
+// Put will return a resource to the pool. For every successful Get,
+// a corresponding Put is required. If you no longer need a resource,
+// you will need to call Put(nil) instead of returning the closed resource.
+// The will eventually cause a new resource to be created in its place.
+func (p *FastPool) Put(resource Resource) {
+	p.Lock()
+
+	p.state.InUse--
+
+	if p.state.Closed || p.active() > p.state.Capacity {
+		// We're ether closed or shrinking.
+		p.Unlock()
+		if resource != nil {
+			resource.Close()
+		}
+		return
+	}
+
+	if resource == nil {
+		p.Unlock()
+		p.ensureMinimumActive()
```

Comment:
Can this block?

Reply:
Yes it can! Good find.

I'm thinking of converting `ensureMinimumActive` to goroutine, similar to `closeIdleResources()`.